Fix problem handling annotated types on type use - continued

nfekete opened the pull request:
The previous commit that supposed to fix the problem with annotated type
use under javac solved only some of the possible cases. Upon further
use, I faced another problem when an annotated type use was combined
with a type that was outside of bindgen's scope. In that case a
GenericObjectBinding was generated for the type and the problem
described in https://github.com/igloo-project/bindgen/issues/8 came back
in a slightly different form. I decided I'll try to do a better job this
time, so I decided to replace all invocations of TypeMirror.toString
with Util.getTypeName. Hopefully this will solve the issue completely.

I also added a JUnit test for the two cases I uncovered up to this
point.

nfekete committed:
Fix problem handling annotated types on type use - continued

The previous commit that supposed to fix the problem with annotated type
use under javac solved only some of the possible cases. Upon further
use, I faced another problem when an annotated type use was combined
with a type that was outside of bindgen's scope. In that case a
GenericObjectBinding was generated for the type and the problem
described in https://github.com/igloo-project/bindgen/issues/8 came back
in a slightly different form. I decided I'll try to do a better job this
time, so I decided to replace all invocations of TypeMirror.toString
with Util.getTypeName. Hopefully this will solve the issue completely.

I also added a JUnit test for the two cases I uncovered up to this
point.

nfekete commented:
This is the PR I [mentioned](https://github.com/igloo-project/bindgen/issues/8#issuecomment-444222698) earlier, that will hopefully fix the problem with annotated types on type use completely.

lalmeras commented:
This fix breaks generation on my test-project. Bindings for generic classes located in dependencies are no longer generated (but binding for java.util.SortedMap is generated) . I try to find a minimal test-case.

lalmeras commented:
My last assumption is wrong. This is not related on dependency/not dependency.

I should just check bindgen-exception.txt. My generic type binding cannot be generated, and this failure is silent, except for this indirectly related message:

```
diagnostic: error: Error writing out error message Attempt to reopen a file for path /home/lalmeras/git/fafcea-si-logart/fafcea-si-test/target/generated-sources/apt/bindgen-exception.txt
```
I think the failure is linked to the rare pattern of defining a generic as the intercection of two types.

Here is an example of a broken type:

```
public class GenericEntityReference<K extends Comparable<K> & Serializable, E extends Serializable> {...}
```
I'll add a new test-case so we can work on this. We also need to make this error an explicit one.

lalmeras committed:
Added new test case on multiple bounds generics

lalmeras commented:
Sorry for the delay, I just add a new test-case in 7244494cb608d75759533195e23d17c412c8784c . This use-case works just fine on dev branch, and it can be cherry-picked from dev where it triggers the issue I talk about.

lalmeras committed:
move from SimpleTypeVisitor6 to SimpleTypeVisitor8

SimpleTypeVisitor8 is compatible with IntersectionType.
SimpleTypeVisitor6 just throws an error when used with such a type.

lalmeras commented:
I push test-case and fix on the branch.

I just dig further on my failed use-case ; this is related to an unimplemented visitIntersection(...) in SimpleTypeVisitor6 ; I move to SimpleTypeVisitor8 (java 8) implementation. As we already depend on lambda, there is no consequence to do this change.

This change fixes the use-case I added. Is it OK for you @nfekete ?

nfekete commented:
Sorry, I was quite busy lately, I haven't got the chance to take a look at this yet. If that simple change in your last commit solves the issue, that's great. Java 8 is required anyway, so yea, I'm totally OK with it :)

Thanks for taking the time to look into it.
